lib/HLSL/DxilContainerReflection.cpp

```diff
@@ -430,7 +438,494 @@ ID3D12ShaderReflectionConstantBuffer *CInvalidSRVariable::GetBuffer() {
+  case hlsl::DXIL::ComponentType::I32:
+    componentType = D3D_SVT_INT;
+    m_Name = "int";
+    break;
+
+  case hlsl::DXIL::ComponentType::U64:
+#ifdef DBG
+    OutputDebugStringA("DxilContainerReflection.cpp: warning: component of type 'U64' being reflected as if 'U32'\n");
+#endif
+  case hlsl::DXIL::ComponentType::U32:
+    componentType = D3D_SVT_UINT;
+    m_Name = "uint";
+    break;
+
+  case hlsl::DXIL::ComponentType::F16:
+  case hlsl::DXIL::ComponentType::SNormF16:
+  case hlsl::DXIL::ComponentType::UNormF16:
+    componentType = D3D_SVT_MIN16FLOAT;
+    m_Name = "min16float";
+    break;
+
+  case hlsl::DXIL::ComponentType::F32:
+  case hlsl::DXIL::ComponentType::SNormF32:
+  case hlsl::DXIL::ComponentType::UNormF32:
+    componentType = D3D_SVT_FLOAT;
+    m_Name = "float";
+    break;
+
+  case hlsl::DXIL::ComponentType::F64:
+  case hlsl::DXIL::ComponentType::SNormF64:
+  case hlsl::DXIL::ComponentType::UNormF64:
+    componentType = D3D_SVT_DOUBLE;
+    m_Name = "double";
+    break;
+
+  default:
+#ifdef DBG
+    OutputDebugStringA("DxilContainerReflection.cpp: error: unknown component type\n");
+#endif
+    break;
+  }
+  m_Desc.Type = componentType;
+
+  // A matrix type is encoded as a vector type, plus annotations, so we
+  // need to check for this case before other vector cases.
+  if(typeAnnotation.HasMatrixAnnotation())
+  {
+    // We can extract the details from the annotation.
+    DxilMatrixAnnotation const& matrixAnnotation = typeAnnotation.GetMatrixAnnotation();
+
+    switch(matrixAnnotation.Orientation)
+    {
+    default:
+#ifdef DBG
+      OutputDebugStringA("DxilContainerReflection.cpp: error: unknown matrix orientation\n");
+#endif
+    // Note: column-major layout is the default
+    case hlsl::MatrixOrientation::Undefined:
+    case hlsl::MatrixOrientation::ColumnMajor:
+      m_Desc.Class = D3D_SVC_MATRIX_COLUMNS;
+      break;
+
+    case hlsl::MatrixOrientation::RowMajor:
+      m_Desc.Class = D3D_SVC_MATRIX_ROWS;
+      break;
+    }
+
+    m_Desc.Rows = matrixAnnotation.Rows;
+    m_Desc.Columns = matrixAnnotation.Cols;
+    m_Name += std::to_string(matrixAnnotation.Rows) + "x" + std::to_string(matrixAnnotation.Cols);
+  }
+  else if( type->isVectorTy() )
+  {
+    // We assume that LLVM vectors either represent matrices (handled above)
+    // or HLSL vectors.
+    //
+    // Note: the reflection interface encodes an N-vector as if it had 1 row
+    // and N columns.
+    m_Desc.Class = D3D_SVC_VECTOR;
+    m_Desc.Rows = 1;
+    m_Desc.Columns = type->getVectorNumElements();
+
+    m_Name += std::to_string(type->getVectorNumElements());
+  }
+  else if( type->isStructTy() )
+  {
+    // A struct type might be an ordinary user-defined `struct`,
+    // or one of the builtin in HLSL "object" types.
+    StructType *structType = cast<StructType>(type);
+
+    // We use our function to try to detect an object type
+    // based on its name.
+    if(TryToDetectObjectType(structType, &m_Desc.Type))
```

Comment:
Great. I'll remove TryToDetectObjectType after finish the pass to strip object types from cbuffer.